compiler/next/include/chpl/AST/BlockStmt.h

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright 2021 Hewlett Packard Enterprise Development LP
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef CHPL_AST_BLOCKSTMT_H
+#define CHPL_AST_BLOCKSTMT_H
+
+#include "chpl/AST/Location.h"
+#include "chpl/AST/Stmt.h"
+
+namespace chpl {
+namespace ast {
+
+/**
+  This class represents a { } block.
+ */
+class BlockStmt final : public Stmt {
+ private:
+  BlockStmt(ASTList stmts);
+  bool contentsMatchInner(const BaseAST* other) const override;
+
+ public:
+  /**
+   Create and return a BlockStmt containing the passed stmts.
+   */
+  static owned<BlockStmt> build(Builder* builder, Location loc, ASTList stmts);
+
+  int numStmts() const {
+    return this->numChildren();
+  }
+  const Expr* stmt(int i) const {
```

Comment:
I think so, but do you have another name suggestion you would like to make?

Since the AST classes are mostly immutable, and as a matter of style anyway, I prefer something like `property()` to `getProperty()`. Relatedly, I'd rather call it `stmt(int i)` rather than `getStmt(int i)`.

But if you think we shouldn't name it `stmt` but rather something else I'm open to it. I think `stmt` is reasonable because I view BlockStmt as basically wrapping a list of statements, so `stmt(i)` is a reasonable name. Would you rather have it be called `statement(int i)` ?

Reply:
It seems like "expr" would be more appropriate since it is returning an Expr?

Reply:
Oh, right. I think it's returning a "statement" in the sense of the language and the parser. But we can have things like `a;` which are going to be `Expr*` in the AST (unless we change approach). IMO the distinction between statements and exprs isn't particularly useful after parsing.

I don't like the name `expr` for the things within a Block because that would have me think of the the i'th expression in a postorder traversal (say).

We could have a different name though that is neither `stmt()` or `expr()`. Do you have a suggestion?

In terms of changing approach - we could make `Expr` inherit from `Stmt` instead of the other way around (`clang` does this but it seems really weird to me) or we could create an AST node just to represent an Expr that is at the statement level.

Reply:
I don't think it's necessary to add a node class just for this case since it doesn't seem like it has much impact, but maybe there's something that can be done to encompass that meaning in the name.  Would `stmtExpr()` as a name help or would it just feel like nonsense?  I think I'd also be okay with including the explanation in a comment on the function since we probably won't change that part of the hierarchy after we get past a certain point

Reply:
I've added a comment
```
  /**
   Return the i'th statement in the block. Note that since simple expressions
   are also statements, this function returns an Expr.
   */
```

Reply:
I think it will be easy to get used to this style of names, especially if it is followed consistently, like for accessing i-th formal, i-th zippered iterable, i-th cobegin participant, ... Newcomers may have a learning curve for this, however I expect it to be insignificant.

Put another way, I wish I had a better proposal, I don't, and I think it's OK.

Reply:
Just to tie things up a bit - this comment thread is mostly resolved by not having a Stmt base class.